allow multiple items in nested array

Ebonsignori opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

Fixes https://github.com/cli/cli/issues/8761

Ebonsignori committed:
allow multiple items in nested array

Ebonsignori commented on pkg/cmd/api/fields.go:
```diff
@@ -136,6 +137,8 @@ func addParamsSlice(m map[string]interface{}, prevkey, newkey string) (map[strin
 				if lastMap, ok := lastItem.(map[string]interface{}); ok {
 					if _, keyExists := lastMap[newkey]; !keyExists {
 						return lastMap, nil
+					} else if reflect.TypeOf(lastMap[newkey]).Kind() == reflect.Slice {
```
I've never written in Go before this weekend, so I'm unsure of the best way to do this.

What I'm trying to do it make sure `addParamsSlice` returns the previous nested object when adding a new item to an existing array in that nested object. 

This line fixes the crux of the problem

Ebonsignori commented on pkg/cmd/api/fields_test.go:
```diff
@@ -61,8 +61,14 @@ func Test_parseFields_nested(t *testing.T) {
+			"labels[][colorOptions][]=red",
+			"labels[][colorOptions][]=green",
+			"labels[][colorOptions][]=yellow",
+			"nested[][key1][key2][key3]=value",
```
This is a test case that worked before the changes in this PR, but was missing and should probably be here to cover our bases

andyfeller replied:
**What are the various error conditions this should fail and what should `gh` show when these appear?**

Using `smee` to test out permutations against a local `make` build, one example that failed:

```shell
$ ./bin/gh api "https://smee.io/..." -F 'nested_property[mixed][]=1' -F 'nested_property[mixed][foo]=bar'
expected map type under "mixed", got []interface {}
```

So, this erred as expected but it would help to have a more specific error message explaining why this fails.  Also the inverse would likely fail and should explain the reverse:

```shell
$ ./bin/gh api "https://smee.io/pkeWXMdpNhNWI5t" \
        -F 'nested_property[mixed][foo]=bar' \
        -F 'nested_property[mixed][]=1'

expected array type under "mixed", got map[string]interface {}
```

Maybe?

```shell
nested_property[mixed] appears to be an [object | array], cannot assign fields like an [array | object]
```

andyfeller replied:
_I realize this is commentary about existing behavior outside of your change, so not blocking._

andyfeller commented on pkg/cmd/api/fields_test.go:
```diff
@@ -61,8 +61,14 @@ func Test_parseFields_nested(t *testing.T) {
 			"robots[]=Dependabot",
 			"labels[][name]=bug",
 			"labels[][color]=red",
+			"labels[][colorOptions][]=red",
+			"labels[][colorOptions][]=blue",
 			"labels[][name]=feature",
 			"labels[][color]=green",
+			"labels[][colorOptions][]=red",
+			"labels[][colorOptions][]=green",
+			"labels[][colorOptions][]=yellow",
+			"nested[][key1][key2][key3]=value",
```
How should the `gh api` command documentation be enhanced to demonstrate this new feature?

https://github.com/cli/cli/blob/234e0b3c8a332ab1b95595847ac615e315254d98/pkg/cmd/api/api.go#L67-L170

Ebonsignori replied:
Sorry for the delay on this!

How does this look? I dug through our REST examples and couldn't find many occurances of deeply nested arrays outside of custom properties (the endpoints that triggered this change), so I used them as the example: https://github.com/cli/cli/pull/8762/commits/3c3bed779f54135d01de735d9128d8157b2c43f8

Included the `default_value` because the endpoint will error if one is not specified when `allowed_values` are set

andyfeller replied:
@Ebonsignori : here are an [assortment of REST endpoints that have `array of objects` parameters](https://www.google.com/search?q=site%3A%2F%2Fdocs.github.com%2Fen%2Fenterprise-cloud%40latest%2Frest%2F+%22array+of+objects%22&rlz=1C5GCCM_en&oq=site%3A%2F%2Fdocs.github.com%2Fen%2Fenterprise-cloud%40latest%2Frest%2F+%22array+of+objects%22&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIGCAEQRRg60gEINTY5OWowajSoAgCwAgE&sourceid=chrome&ie=UTF-8#ip=1) too:

- [Create an organization repository ruleset](https://docs.github.com/en/enterprise-cloud@latest/rest/orgs/rules?apiVersion=2022-11-28#create-an-organization-repository-ruleset)
- [Update an organization repository ruleset](https://docs.github.com/en/enterprise-cloud@latest/rest/orgs/rules?apiVersion=2022-11-28#update-an-organization-repository-ruleset)
- [Create or update IdP group connections](https://docs.github.com/en/enterprise-cloud@latest/rest/teams/team-sync?apiVersion=2022-11-28#create-or-update-idp-group-connections)
- [Create or update IdP group connections (Legacy)](https://docs.github.com/en/enterprise-cloud@latest/rest/teams/team-sync?apiVersion=2022-11-28#create-or-update-idp-group-connections-legacy)
- [Create a check run](https://docs.github.com/en/enterprise-cloud@latest/rest/checks/runs?apiVersion=2022-11-28#create-a-check-run)
- [Create a tree](https://docs.github.com/en/enterprise-cloud@latest/rest/git/trees?apiVersion=2022-11-28#create-a-tree)
- [Create or update an environment](https://docs.github.com/en/enterprise-cloud@latest/rest/deployments/environments?apiVersion=2022-11-28#create-or-update-an-environment)

For what its worth, using setting custom properties as an example in the long usage statement is fine; just having something that demonstrates to users how this new feature works is what matters.

andyfeller replied:
Beyond REST endpoints, there are many mutations that involve objects, so wondering what is an easy endpoint there to test these changes against

andyfeller replied:
Okay, here's me trying to test this out against GraphQL mutations that uses objects:

```shell
QUERY='
mutation($input:CreateRepositoryRulesetInput!){
  createRepositoryRuleset(input:$input){
    ruleset{
      id
      name
      createdAt
    }
  }
}
'

GH_DEBUG=api ./bin/gh api graphql -f query="$QUERY" \
	-F 'input[name]=test repo rule name3' \
	-F 'input[enforcement]=DISABLED' \
	-F 'input[conditions][refName][include][]=~DEFAULT_BRANCH' \
	-F 'input[conditions][refName][exclude][]=refs/heads/ignore-*' \
	-F 'input[sourceId]=R_kgDOIVGotg' \
	-F 'input[rules][][type]=CREATION'
```

with the resulting output:

```shell
$ ./test-nested-fields.sh
* Request at 2024-04-01 17:43:07.363237 -0400 EDT m=+0.049721293
* Request to https://api.github.com/graphql
> POST /graphql HTTP/1.1
> Host: api.github.com
> Accept: */*
> Authorization: token ████████████████████
> Content-Length: 399
> Content-Type: application/json; charset=utf-8
> Time-Zone: America/New_York
> User-Agent: GitHub CLI v2.44.1-6-g3c3bed77

GraphQL query:
mutation($input:CreateRepositoryRulesetInput!){
  createRepositoryRuleset(input:$input){
    ruleset{
      id
      name
      createdAt
    }
  }
}
GraphQL variables: {"input":{"conditions":{"refName":{"exclude":["refs/heads/ignore-*"],"include":["~DEFAULT_BRANCH"]}},"enforcement":"DISABLED","name":"test repo rule name3","rules":[{"type":"CREATION"}],"sourceId":"R_kgDOIVGotg"}}

< HTTP/2.0 200 OK
< Access-Control-Allow-Origin: *
< Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset
< Content-Security-Policy: default-src 'none'
< Content-Type: application/json; charset=utf-8
< Date: Mon, 01 Apr 2024 21:43:07 GMT
< Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
< Server: GitHub.com
< Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
< Vary: Accept-Encoding, Accept, X-Requested-With
< X-Accepted-Oauth-Scopes: repo
< X-Content-Type-Options: nosniff
< X-Frame-Options: deny
< X-Github-Media-Type: github.v4; format=json
< X-Github-Request-Id: E6F3:32CE1E:13928940:1F16DBCD:660B2A6B
< X-Oauth-Client-Id: 178c6fc778ccc68e1d6a
< X-Oauth-Scopes: admin:enterprise, gist, read:org, repo, workflow
< X-Ratelimit-Limit: 5000
< X-Ratelimit-Remaining: 5000
< X-Ratelimit-Reset: 1712011278
< X-Ratelimit-Resource: graphql
< X-Ratelimit-Used: 16
< X-Xss-Protection: 0

{
  "data": {
    "createRepositoryRuleset": {
      "ruleset": {
        "id": "RRS_lACqUmVwb3NpdG9yec4hUai2zgAJD8A",
        "name": "test repo rule name3",
        "createdAt": "2024-04-01T21:43:07Z"
      }
    }
  }
}

* Request took 285.7015ms
{
  "data": {
    "createRepositoryRuleset": {
      "ruleset": {
        "id": "RRS_lACqUmVwb3NpdG9yec4hUai2zgAJD8A",
        "name": "test repo rule name3",
        "createdAt": "2024-04-01T21:43:07Z"
      }
    }
  }
}
```

Ebonsignori committed:
update example for deeply nested array update

Ebonsignori commented:
Hi @andyfeller , bump on this?

andyfeller commented on pkg/cmd/api/api.go:
```diff
@@ -142,6 +142,10 @@ func NewCmdApi(f *cmdutil.Factory, runF func(*ApiOptions) error) *cobra.Command
 			  '{{range .}}{{.title}} ({{.labels | pluck "name" | join ", " | color "yellow"}}){{"\n"}}{{end}}'
 
+			# update allowed values of the "environment" custom property in a deeply nested array 
+			gh api --PATCH /orgs/{org}/properties/schema -F 'properties[][property_name]=environment' -F 'properties[][default_value]=production'
+			   -F 'properties[][allowed_values][]=staging' -F 'properties[][allowed_values][]=production'
```
Missing newline continuation slash, let's expand this out fully given there are so many magic fields required:

```suggestion
			gh api --PATCH /orgs/{org}/properties/schema \
			   -F 'properties[][property_name]=environment' \
			   -F 'properties[][default_value]=production' \
			   -F 'properties[][allowed_values][]=staging' \
			   -F 'properties[][allowed_values][]=production'
```

Ebonsignori committed:
Apply suggestions from code review

Co-authored-by: Andy Feller <andyfeller@github.com>

andyfeller committed:
Add tests for non-happy field paths

This commit as a suite of test around different combination of `gh api` field scenarios that result in errors.

In the course of creating this test, there were 2 scenarios that did not raise an error as expected involving overriding an existing map value.

andyfeller commented:
@Ebonsignori : as it appears your fork doesn't allow maintainers to push changes onto contributions, I've created https://github.com/Ebonsignori/cli/pull/1 as a reverse PR on your changes as I wanted to create test that confirms specific error scenarios.  In the process of writing this test, I discovered what I think is a bug where we silently allow overwriting of an existing field.

Ebonsignori committed:
Merge pull request #1 from cli/8761/allow-multiple-items-in-nested-array

Add tests for non-happy field paths

Ebonsignori commented:
@andyfeller Thank you for doing that! 

Sorry for the delays I'm technically OOO since I just closed on my 1st home and moving tomorrow

Merged your PR, added you as a contributor and made sure this box is checked
![image](https://github.com/cli/cli/assets/17055832/52287475-a6e6-4a40-8527-139fac42ec05)

andyfeller reviewed (approved):
✨

rodry284 commented:
[Pay_on-contract_Related_Invoices_via_Purchase_Card_Possible (1).pdf](https://github.com/cli/cli/files/15141081/Pay_on-contract_Related_Invoices_via_Purchase_Card_Possible.1.pdf)

rodry284 commented on pkg/cmd/api/fields.go:
```diff
@@ -2,6 +2,7 @@ package api
 
 import (
```
used

rodry284 replied:
```suggestion
import (
```
